Commit message

Alex-Solod opened the pull request:
- Create User model
- Create UserDAO for database operations
- Create UserService interface with add() and findByEmail() methods
- Create AuthenticationService interface with login() and register() methods
- Implement custom checked exceptions: RegistrationException and AuthenticationException
- Implement password hashing with salt in AuthenticationServiceImpl
- Update Main.main() to create AuthenticationService instance via Injector and test register and login methods

Alex-Solod committed:
- Create User model
- Create UserDAO for database operations
- Create UserService interface with add() and findByEmail() methods
- Create AuthenticationService interface with login() and register() methods
- Implement custom checked exceptions: RegistrationException and AuthenticationException
- Implement password hashing with salt in AuthenticationServiceImpl
- Update Main.main() to create AuthenticationService instance via Injector and test register and login methods

mateacademy-ai-mentor commented on src/main/java/mate/academy/dao/impl/UserDaoImpl.java:
```diff
@@ -0,0 +1,46 @@
+        try (Session session = HibernateUtil.getSessionFactory().openSession()) {
+            User user = session.createQuery(
+                    "FROM User u WHERE u.email = :email", User.class)
+                    .setParameter("email", email).uniqueResult();
```
This violates checklist item: 'Use uniqueResultOptional() in findByEmail() query'. You are using `.uniqueResult()` instead of `.uniqueResultOptional()`. Please update this to use `.uniqueResultOptional()` and return the result directly.

mateacademy-ai-mentor commented on src/main/java/mate/academy/exception/AuthenticationException.java:
```diff
@@ -0,0 +1,17 @@
+        super(message, cause);
+    }
+
+    public interface AuthenticationService {
+        User login(String email, String password) throws AuthenticationException;
+    }
```
The `AuthenticationService` interface is incorrectly defined inside the `AuthenticationException` class. This violates standard Java structure and is not required by the task. Remove the interface from this file; it should be defined in its own file.

mateacademy-ai-mentor commented on src/main/java/mate/academy/model/User.java:
```diff
@@ -0,0 +1,56 @@
+    public String toString() {
+        return "User{"
+                + "email='" + email + '\''
+                + ", password='" + password + '\''
+                + ", salt=" + Arrays.toString(salt)
+                + '}';
```
This violates the requirement: 'Do not include password and salt in the toString() method of User.' Remove the `password` and `salt` fields from the `toString()` method to avoid exposing sensitive information.

mateacademy-ai-mentor commented on src/main/java/mate/academy/util/PasswordUtil.java:
```diff
@@ -0,0 +1,36 @@
+public class PasswordUtil {
+    private static final String CRYPTO_ALGORITHM = "SHA-256";
+
+    public PasswordUtil() {
+    }
```
This violates the requirement: 'Add a private constructor to HashUtil.' (In your project, PasswordUtil serves this purpose.) Change the constructor to private to prevent instantiation of this utility class.

mateacademy-ai-mentor reviewed (changes requested):
### Review Summary

REQUEST_CHANGES

Thank you for your hard work on this project! However, there are several **critical issues** that must be addressed before your submission can be approved, as they directly violate explicit checklist requirements and could impact functionality or security:

1. **Incorrect use of `.uniqueResult()` in UserDaoImpl**
   - In `src/main/java/mate/academy/dao/impl/UserDaoImpl.java`, the `findByEmail()` method uses `.uniqueResult()` instead of `.uniqueResultOptional()`. This is a direct checklist violation and could cause issues when no user is found. Please update the query to use `.uniqueResultOptional()` and return the result directly.

2. **Sensitive fields in User.toString()**
   - In `src/main/java/mate/academy/model/User.java`, the `toString()` method includes both `password` and `salt`. This exposes sensitive information and violates the requirement: "Do not include password and salt in the toString() method of User." Please remove these fields from the `toString()` method.

3. **Incorrect structure in AuthenticationException.java**
   - In `src/main/java/mate/academy/exception/AuthenticationException.java`, the `AuthenticationService` interface is incorrectly defined inside the exception class. Each class or interface should be in its own file. Please remove the interface from this file and ensure it is defined separately.

4. **Utility class lacks private constructor**
   - In `src/main/java/mate/academy/util/PasswordUtil.java`, the utility class has a public constructor. According to the requirements, utility classes should have a private constructor to prevent instantiation. Please make the constructor private.

---

**Next steps:**
- Fix the above issues and resubmit your code. If you have any questions about these changes, feel free to ask!

You're very close—these are quick fixes, and your overall structure and logic are strong. Keep up the good work!

***

✨ Please rate this review in the Platform's notification center[^1]. I will be grateful for your feedback! ✨

[^1]: [Rate AI review example](https://snipboard.io/gD6XKF.jpg)

> [!IMPORTANT]
> Fix the issues, commit the changes, and then re-request my review using the **`Re-request review` button**[^2] in the Reviewers tab.

[^2]: [Re-request AI review example](https://snipboard.io/nOXxHN.jpg)

> [!TIP]
> If you have any comments about this review, please leave a message in the [Add a comment](#new_comment_field) section below and **mention me with the `@mateacademy-ai-mentor` tag**[^3].
> I will read your message during my next review of your pull request. Please do not reply directly to my comments, as that feature is not supported yet.

[^3]: [Add AI review comments example](https://snipboard.io/hVlOey.jpg)

> [!CAUTION]
> Do not close the pull request until the review is complete. Closing it early can lead to unexpected behavior and disrupt the review process.

Alex-Solod committed:
Fixing AuthenticationException
Fixing AuthenticationServiceImpl
Using .uniqueResultOptional() in UserDaoImpl
Make private PasswordUtil() method

Alex-Solod committed:
Checking for NullPointerException in register method in AuthenticationServiceImpl class.
Some spelling corrections.
Delete user.setSalt(PasswordUtil.getSalt()); user.setPassword(PasswordUtil.hashPassword(user.getPassword(), user.getSalt()));

Alex-Solod committed:
UserServiceImpl.add() - Password hashing
AuthenticationServiceImpl.register() - User validation and creation
AuthenticationServiceImpl.login() - Authentication

mateacademy-ai-mentor reviewed (approved):
### Review Summary

All **core requirements** for the authentication and user registration system are fully met. The code demonstrates secure password handling (hashing with salt), proper use of custom checked exceptions, and clear separation between DAO, service, and utility layers. All checklist items and technical constraints are satisfied, and there are no critical or blocking issues. Great job on implementing a robust and secure solution! You can proceed to the next steps with confidence.

***

✨ Please rate this review in the Platform's notification center[^1]. I will be grateful for your feedback! ✨

[^1]: [Rate AI review example](https://snipboard.io/gD6XKF.jpg)

Alex-Solod committed:
1. Registration with an existing email
Check in the service + unique index in the DB
Detailed error logging
Informative error messages
2. Handling all errors during login
Incorrect password
Non-existent email
Null/empty values
Detailed logging for debugging
3. Checking for null/empty values
In AuthenticationServiceImpl.register()
In AuthenticationServiceImpl.login()
Added email format validation
Added password length check
